File: julia/p2.jl
function eval_password(line)
    parts = split(line, " ");
    idxs = split(parts[1], "-");
    _min = parse(Int, idxs[1]);
    _max = parse(Int, idxs[2]);
    letter = parts[2][1];
    password = parts[3];
    cnt = 0;

    for c in password
        if c == letter
            cnt += 1;
        end
    end

    return (_min <= cnt) & (cnt <= _max)
end

function eval_password2(line)
    parts = split(line, " ");
    idxs = split(parts[1], "-");
    _min = parse(Int, idxs[1]);
    _max = parse(Int, idxs[2]);
    letter = parts[2][1];
    password = parts[3];

    return ((password[_min] == letter) & (password[_max] != letter)) |
           ((password[_min] != letter) & (password[_max] == letter))
end

function main()
    fname = "../data/p2.txt";
    open(fname) do f
        tot1 = 0;
        tot2 = 0;
        for (i, line) in enumerate(eachline(f))
            tot1 = tot1 + eval_password(line);
            tot2 = tot2 + eval_password2(line);
        end
        println("Tot part 1: ", tot1);
        println("Tot part 2: ", tot2);
    end
end

main()

File: python/p2.py
def eval_password(line):
    tokens = line.split(' ')
    [_min, _max] = tokens[0].split('-')
    _min = int(_min)
    _max = int(_max)
    letter = tokens[1][0]
    password = tokens[2]
    cnt = 0

    for c in password:
        if c == letter:
            cnt += 1
    return _min <= cnt and cnt <= _max

def eval_password2(line):
    tokens = line.split(' ')
    [_min, _max] = tokens[0].split('-')
    _min = int(_min) - 1
    _max = int(_max) - 1
    letter = tokens[1][0]
    password = tokens[2]

    return (password[_min] == letter and password[_max] != letter) or \
           (password[_min] != letter and password[_max] == letter)

def main():
    f = open('../data/p2.txt', 'r')
    tot = 0

    for line in f.readlines():
        tot += eval_password2(line)
    print(tot)

if __name__ == '__main__':
    main()
